### PR TITLE
Add a document for leveraging Advanced Matrix Extensions

### LINKED ISSUE #2355: 💡 [REQUEST] - Write a tutorial about how to leverage AMX with PyTorch on the 4th Gen of Xeon
### 🚀 Descirbe the improvement or the new tutorial

The 4th Generation Intel® Xeon® Scalable Processor platform is an unique, scalable platform optimized for different workloads acceleration on AI. The new built-in AI acceleration engine, Intel® Advanced Matrix Extensions (AMX) is able to accelerate a variety of AI Inference and Training workloads (NLP, recommendation systems, image recognition…) with BF16 and INT8 datatype.

PyTorch has enabled AMX support for computation intensive operators, e.g. Conv2d, ConvTranspose2d, Linear, MatMul, bmm with `torch.bfloat16` datatype and int8 on the quantization backend. It is better to write a tutorial to tell users how to leverage AMX on PyTorch.

### Existing tutorials on this topic

_No response_

### Additional context

We aim to complete the document as part of PyTorch Docathon 2023. cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ZailiWang @ZhaoqiongZ @leslie-fang-intel @Xia-Weiwen @sekahler2 @CaoE @zhuhaozhe @Valentine233 @caoe

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #2355

## Description

Add a document about how to leverage AMX with PyTorch on the 4th Gen of Xeon.

## Checklist

- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ZailiWang @ZhaoqiongZ @leslie-fang-intel @Xia-Weiwen @sekahler2 @zhuhaozhe @Valentine233

### CODE REVIEW

File: recipes_source/amx.rst [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``linear``
```
Comment (jgong5) [inline recipes_source/amx.rst:52]: I guess we need a special note for quantized linear here that whether AMX kernel is chosen also depends on the policy of the quantization backend. Currently, the x86 quant backend uses fbgemm, not onednn while users can use onednn backend to turn on AMX for linear op. cc @Xia-Weiwen

In general, it is also true that whether to dispatch to AMX kernels is a backend/library choice. The backend/library would choose the most optimal kernels. It is worth noting in this tutorial.

Comment (Xia-Weiwen) [reply recipes_source/amx.rst:52]: Yes. However, I am not sure if it's OK to give such details in tutorial. 🤔

File: recipes_source/amx.rst [hunk 2]
```diff
@@ -0,0 +1,108 @@
+==============================================
+Leverage Advanced Matrix Extensions
+==============================================
+
+Introduction
+============
+
+Advanced Matrix Extensions (AMX), also known as Intel® Advanced Matrix Extensions (Intel® AMX), is an extension to the x86 instruction set architecture (ISA).
+Intel advances AI capabilities with 4th Gen Intel® Xeon® Scalable processors and Intel® AMX, delivering 3x to 10x higher inference and training performance versus the previous generation, see `Accelerate AI Workloads with Intel® AMX`_.
+AMX supports two data types, INT8 and BFloat16, compared to AVX512 FP32, it can achieve up to 32x and 16x acceleration, respectively, see figure 6 of `Accelerate AI Workloads with Intel® AMX`_.
```
Comment (msaroufim) [inline recipes_source/amx.rst:10]: Are the speedups only on some particular newer hardware? Is the hardware consumer or enterprise centric?

File: recipes_source/amx.rst [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+- torch.compile:
+
+When the generated graph model runs into oneDNN implementations with the supported operators mentioned in lists above, AMX accelerations will be activated.
+
+
+Confirm AMX is being utilized
+------------------------------
+
+Set environment variable ``export ONEDNN_VERBOSE=1`` to get oneDNN verbose at runtime.
```
Comment (msaroufim) [inline recipes_source/amx.rst:84]: it would be nice to have some python function like `is_x_available()`

File: recipes_source/amx.rst [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+``conv3d``,
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``linear``
+
+Note: For quantized linear, whether to leverage AMX depends on which quantization backend to choose.
+At present, x86 quantization backend is used by default for quantized linear, using fbgemm, while users can specify onednn backend to turn on AMX for quantized linear.
+
+Guidelines of leveraging AMX with workloads
```
Comment (msaroufim) [inline recipes_source/amx.rst:57]: I would start with this section on how to use it and have the supported ops show up at the bottm

File: recipes_source/amx.rst [hunk 5]
```diff
@@ -0,0 +1,108 @@
+==============================================
+Leverage Advanced Matrix Extensions
+==============================================
+
+Introduction
+============
+
+Advanced Matrix Extensions (AMX), also known as Intel® Advanced Matrix Extensions (Intel® AMX), is an extension to the x86 instruction set architecture (ISA).
```
Comment (msaroufim) [inline recipes_source/amx.rst:8]: I realize AMX is lower level than other Intel technologies but it's still worth rationalizing to an end user in a few lines why it's interesting for them to know about AMX vs Intel compiler technologies

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: recipes_source/amx.rst (hunk 4 above)
Comment (CaoE) [reply recipes_source/amx.rst:56]: Thanks for your suggestion. Do you mean to move this section above the supported ops ? Like this：
AMX in PyTorch
   Guidelines of leveraging AMX with workloads
   List supported ops
   ...

File: recipes_source/amx.rst (hunk 2 above)
Comment (mingfeima) [reply recipes_source/amx.rst:10]: AMX is only available from the 4th gen of Xeon (codename  sapphire rapids), it is enterprise centric.

File: recipes_source/amx.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+Intel advances AI capabilities with 4th Gen Intel® Xeon® Scalable processors and Intel® AMX, delivering 3x to 10x higher inference and training performance versus the previous generation, see `Accelerate AI Workloads with Intel® AMX`_.
+AMX supports two data types, INT8 and BFloat16, compared to AVX512 FP32, it can achieve up to 32x and 16x acceleration, respectively, see figure 6 of `Accelerate AI Workloads with Intel® AMX`_.
```
Comment (mingfeima) [inline recipes_source/amx.rst:10]: we can directly copy the wording from

```
Compared to 3rd Gen Intel Xeon Scalable processors running Intel® Advanced Vector Extensions 512 Neural Network Instructions (Intel® AVX-512 VNNI), 4th Gen Intel Xeon Scalable processors running Intel AMX can perform 2,048 INT8 operations per cycle, rather than 256 INT8 operations per cycle. They can also perform 1,024 BF16 operations per cycle, as compared to 64 FP32 operations per cycle.
```

which is a quote from https://www.intel.com/content/www/us/en/products/docs/accelerator-engines/advanced-matrix-extensions/ai-solution-brief.html

File: recipes_source/amx.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+``linear``,
+``matmul``,
+``_convolution``
+
+- Quantization CPU ops that can leverage AMX:
+
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv1d``,
```
Comment (mingfeima) [inline recipes_source/amx.rst:46]: why we have 2 sets of `conv1d`, `conv2d`, `conv3d` here

File: recipes_source/amx.rst [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+``conv_transpose2d``,
+``conv_transpose3d``,
+``bmm``,
+``mm``,
+``baddbmm``,
+``addmm``,
+``addbmm``,
+``linear``,
+``matmul``,
+``_convolution``
```
Comment (mingfeima) [inline recipes_source/amx.rst:39]: `_convolution` is not intended to be directly used, start with a `_`

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: recipes_source/amx.rst (hunk 3 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Add a python function `torch.backends.mkldnn.verbose`.

File: recipes_source/amx.rst (hunk 6 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Quoted this.

File: recipes_source/amx.rst (hunk 7 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed typos.

File: recipes_source/amx.rst (hunk 8 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Removed `_convolution `.

File: recipes_source/amx.rst (hunk 5 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Added more introduction to AMX and the benefits it can bring.

File: recipes_source/amx.rst [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``linear``
+
+Note: For quantized linear, whether to leverage AMX depends on the policy of the quantization backend.
```
Comment (jgong5) [inline recipes_source/amx.rst:81]: Suggest to move this note to `Guidelines of leveraging AMX with workloads` and make it more general like:
```
Note: When you use PyTorch on a CPU that supports AMX, the framework will automatically enable AMX usage by default. This means that PyTorch will attempt to leverage the AMX feature whenever possible to speed up matrix multiplication operations. However, it's important to note that the decision to dispatch to the AMX kernel ultimately depends on the internal optimization strategy of the oneDNN library and the quantization backend, which PyTorch relies on for performance enhancements.
```

Comment (jgong5) [reply recipes_source/amx.rst:81]: Also, you may consider to add the following caveat: "It's worth noting that the specific details of how AMX utilization is handled internally by PyTorch and the oneDNN library may be subject to change with updates and improvements to the framework."

Comment (CaoE) [reply recipes_source/amx.rst]: Thanks for your suggestions. Applied the changes.

**Review by jgong5 — APPROVED** [review]
Please add a "summary" or "conclusion" section to summarize the document.

**Comment by CaoE** [comment]
> Please add a "summary" or "conclusion" section to summarize the document.
Added conclusion section.

**Comment by CaoE** [comment]
@msaroufim Could you please review this doc ? Thanks.

**Comment by CaoE** [comment]
@ngimel Could you please review this doc ? Thank you.

**Comment by CaoE** [comment]
@msaroufim Could you please review this doc ? Thank you.

**Comment by CaoE** [comment]
@kit1980 Could you please review this doc ? Thank you.

File: recipes_source/amx.rst [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+
+AMX in PyTorch
+==============
+
+PyTorch leverages AMX for computing intensive operators with BFloat16 and quantization with INT8 by its backend oneDNN
+to get higher performance out-of-box on x86 CPUs with AMX support.
+For more detailed information of oneDNN, see `oneDNN`_.
+
+The operation is fully handled by oneDNN according to the execution code path generated. I.e. when a supported operation gets executed into oneDNN implementation on a hardware platform with AMX support, AMX instructions will be invoked automatically inside oneDNN.
```
Comment (svekars) [inline recipes_source/amx.rst:25]:
```suggestion
The operation is fully handled by oneDNN according to the execution code path generated. For example, when a supported operation gets executed into oneDNN implementation on a hardware platform with AMX support, AMX instructions will be invoked automatically inside oneDNN.
```

File: recipes_source/amx.rst [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+Using ``torch.cpu.amp`` or ``torch.autocast("cpu")`` would utilize AMX acceleration for supported operators.
+
+::
+
+   model = model.to(memory_format=torch.channels_last)
+   with torch.cpu.amp.autocast():
+       output = model(input)
+
+Note: Use channels last format to get better performance.
```
Comment (svekars) [inline recipes_source/amx.rst:41]:
```suggestion
.. note:: Use channels' last format to get better performance. 
```

File: recipes_source/amx.rst [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+- Quantization:
+
+Applying quantization would utilize AMX acceleration for supported operators.
+
+- torch.compile:
+
+When the generated graph model runs into oneDNN implementations with the supported operators, AMX accelerations will be activated.
+
+Note: When using PyTorch on CPUs that support AMX, the framework will automatically enable AMX usage by default.
```
Comment (svekars) [inline recipes_source/amx.rst:51]:
```suggestion
..  note:: When using PyTorch on CPUs that support AMX, the framework will automatically enable AMX usage by default.
```

File: recipes_source/amx.rst [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``bmm``,
+``mm``,
+``baddbmm``,
+``addmm``,
+``addbmm``,
+``linear``,
+``matmul``,
```
Comment (svekars) [inline recipes_source/amx.rst:75]:
```suggestion
BF16 CPU ops that can leverage AMX:

- ``conv1d``
- ``conv2d``
- ``conv3d``
- ``conv_transpose1d``
- ``conv_transpose2d``
- ``conv_transpose3d``
- ``bmm``
- ``mm``
- ``baddbmm``
- ``addmm``
- ``addbmm``
- ``linear``
- ``matmul``
```

File: recipes_source/amx.rst [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+- Quantization CPU ops that can leverage AMX:
```
Comment (svekars) [inline recipes_source/amx.rst:77]:
```suggestion
Quantization CPU ops that can leverage AMX:
```

File: recipes_source/amx.rst [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``linear``
```
Comment (svekars) [inline recipes_source/amx.rst:85]:
```suggestion
- ``conv1d``
- ``conv2d``
- ``conv3d``
- ``conv_transpose1d``
- ``conv_transpose2d``
- ``conv_transpose3d``
- ``linear``
```

File: recipes_source/amx.rst [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+
+
+Confirm AMX is being utilized
+------------------------------
+
+Set environment variable ``export ONEDNN_VERBOSE=1``, or use ``torch.backends.mkldnn.verbose`` to flexibly enable oneDNN to dump verbose messages.
```
Comment (svekars) [inline recipes_source/amx.rst:92]:
```suggestion
Set environment variable to ``export ONEDNN_VERBOSE=1`` or use ``torch.backends.mkldnn.verbose`` to flexibly enable oneDNN to dump verbose messages.
```

File: recipes_source/amx.rst [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+   model = model.to(memory_format=torch.channels_last)
+   with torch.cpu.amp.autocast():
+       output = model(input)
+
+Note: Use channels last format to get better performance. 
+
+- Quantization:
+
+Applying quantization would utilize AMX acceleration for supported operators.
```
Comment (svekars) [inline recipes_source/amx.rst:45]:
```suggestion
  - Applying quantization would utilize AMX acceleration for supported operators.
```

**Review by svekars — COMMENTED** [review]
A couple editorial fixes for proper HTML rendering.

File: recipes_source/amx.rst (hunk 16 above)
Comment (sanchitintel) [reply recipes_source/amx.rst:92]: I think `to` should not be added here because the specific environment variable we want to use here is `ONEDNN_VERBOSE`, whose value we set to `1`.

File: recipes_source/amx.rst (hunk 10 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Thanks for your comments.
Fixed.

File: recipes_source/amx.rst (hunk 11 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst (hunk 12 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst (hunk 13 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst (hunk 14 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst (hunk 15 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst (hunk 16 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Thanks for your comments. I will keep the original version for this sentence.

File: recipes_source/amx.rst [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+PyTorch leverages AMX for computing intensive operators with BFloat16 and quantization with INT8 by its backend oneDNN
+to get higher performance out-of-box on x86 CPUs with AMX support.
+For more detailed information of oneDNN, see `oneDNN`_.
+
+The operation is fully handled by oneDNN according to the execution code path generated. I.e. when a supported operation gets executed into oneDNN implementation on a hardware platform with AMX support, AMX instructions will be invoked automatically inside oneDNN.
+Since oneDNN is the default acceleration library for PyTorch CPU, no manual operations are required to enable the AMX support.
+
+Guidelines of leveraging AMX with workloads
+-------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (CaoE) [reply recipes_source/amx.rst:30]: Fixed.

File: recipes_source/amx.rst [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+- BFloat16 data type: 
+
+Using ``torch.cpu.amp`` or ``torch.autocast("cpu")`` would utilize AMX acceleration for supported operators.
+
+::
+
+   model = model.to(memory_format=torch.channels_last)
+   with torch.cpu.amp.autocast():
+       output = model(input)
```
[the remark this answers is not in the document]
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst (hunk 17 above)
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

File: recipes_source/amx.rst [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+
+Note: Use channels last format to get better performance. 
+
+- Quantization:
+
+Applying quantization would utilize AMX acceleration for supported operators.
+
+- torch.compile:
+
+When the generated graph model runs into oneDNN implementations with the supported operators, AMX accelerations will be activated.
```
[the remark this answers is not in the document]
Comment (CaoE) [reply recipes_source/amx.rst]: Fixed

**Review by mingfeima — APPROVED** [review]

**Review by msaroufim — APPROVED** [review]

**Review by svekars — APPROVED** [review]
